# Project 1: SAT & ACT Analysis

### Part 1 Data Cleaning

_By: Evonne Tham_

## Import Libraries

In [1]:
import numpy as np
import pandas as pd

## 2017 Data Import and Cleaning

#### 1. Read In SAT & ACT Data
Read in the sat_2017.csv and act_2017.csv files and assign them to appropriately named pandas dataframes.

In [2]:
sat_2017 = pd.read_csv("../dataset/sat_2017.csv")
act_2017 = pd.read_csv("../dataset/act_2017.csv")

#### 2. Display Data

Print the first 10 rows of each dataframe to your jupyter notebook

In [3]:
sat_2017.head(10)

State Participation  Evidence-Based Reading and Writing  \
0               Alabama            5%                                 593   
1                Alaska           38%                                 547   
2               Arizona           30%                                 563   
3              Arkansas            3%                                 614   
4            California           53%                                 531   
5              Colorado           11%                                 606   
6           Connecticut          100%                                 530   
7              Delaware          100%                                 503   
8  District of Columbia          100%                                 482   
9               Florida           83%                                 520   

   Math  Total  
0   572   1165  
1   533   1080  
2   553   1116  
3   594   1208  
4   524   1055  
5   595   1201  
6   512   1041  
7   492    996  
8   468    950  
9   497   1017

In [4]:
act_2017.head(10)

State Participation  English  Math  Reading  Science  \
0              National           60%     20.3  20.7     21.4     21.0   
1               Alabama          100%     18.9  18.4     19.7     19.4   
2                Alaska           65%     18.7  19.8     20.4     19.9   
3               Arizona           62%     18.6  19.8     20.1     19.8   
4              Arkansas          100%     18.9  19.0     19.7     19.5   
5            California           31%     22.5  22.7     23.1     22.2   
6              Colorado          100%     20.1  20.3     21.2     20.9   
7           Connecticut           31%     25.5  24.6     25.6     24.6   
8              Delaware           18%     24.1  23.4     24.8     23.6   
9  District of Columbia           32%     24.4  23.5     24.9     23.5   

  Composite  
0      21.0  
1      19.2  
2      19.8  
3      19.7  
4      19.4  
5      22.8  
6      20.8  
7      25.2  
8      24.1  
9      24.2

#### 3. Verbally Describe Data

Take your time looking through the data and thoroughly describe the data in the markdown cell below. 

##### Answer:

- The dataframe consisted of the break down of scores according to the test structure (both SAT and ACT) in each state. This also include the participation rates of each state.  
- SAT has a test structure of Evidence-Based Reading and Writing and Math (which is then sum up, displayed in one column in the dataframe), while ACT has English, Math, Reading, and Science, the column Composite shows the averge ACT of each state. 

#### 4a. Does the data look complete? 

##### Answer: 

Yes. However, further exploration in the data is required.

In [5]:
sat_2017.isnull().sum()

State                                 0
Participation                         0
Evidence-Based Reading and Writing    0
Math                                  0
Total                                 0
dtype: int64

In [6]:
act_2017.isnull().sum()

State            0
Participation    0
English          0
Math             0
Reading          0
Science          0
Composite        0
dtype: int64

#### 4b. Are there any obvious issues with the observations?

**What is the minimum *possible* value for each test/subtest? What is the maximum *possible* value?**

Consider comparing any questionable values to the sources of your data:
- [SAT](https://blog.collegevine.com/here-are-the-average-sat-scores-by-state/)
- [ACT](https://blog.prepscholar.com/act-scores-by-state-averages-highs-and-lows)

In [7]:
sat_2017.describe()

Evidence-Based Reading and Writing        Math        Total
count                           51.000000   51.000000    51.000000
mean                           569.117647  547.627451  1126.098039
std                             45.666901   84.909119    92.494812
min                            482.000000   52.000000   950.000000
25%                            533.500000  522.000000  1055.500000
50%                            559.000000  548.000000  1107.000000
75%                            613.000000  599.000000  1212.000000
max                            644.000000  651.000000  1295.000000

In [8]:
act_2017.describe()

English       Math    Reading    Science
count  52.000000  52.000000  52.000000  52.000000
mean   20.919231  21.173077  22.001923  21.040385
std     2.332132   1.963602   2.048672   3.151113
min    16.300000  18.000000  18.100000   2.300000
25%    19.000000  19.400000  20.475000  19.900000
50%    20.550000  20.900000  21.700000  21.150000
75%    23.300000  23.100000  24.125000  22.525000
max    25.500000  25.300000  26.000000  24.900000

##### Answer: 

- min score for SAT: EBR = 200, Math = 200, Total = 400
- max score for SAT: EBR = 800, Math = 800, Total = 1600

Since the minimum score for SAT are 200, there might be some issue as the minimum score in the dataset for Math subject is 52. 

- min score for ACT: 1 for English, Math, Reading, Science, Composite (Avg of E, M, R, S scores)
- max score for ACT: 36 for English, Math, Reading, Science, Composite

Given that the mean score for science in ACT is 21.04 and a std of 3.15, a minimum value of 2.3 seem to be quite off. 

There is an addition row in act_2017 which consist of the National Average 

Reference:
- https://blog.prepscholar.com/how-is-the-sat-scored-scoring-charts
- https://blog.prepscholar.com/how-is-the-act-scored#:~:text=The%20ACT%20has%20four%20sections,ranges%20between%201%20and%2036.

#### 4c. Fix any errors you identified

**The data is available** so there's no need to guess or calculate anything. If you didn't find any errors, continue to the next step.

In [9]:
# Drop row that have a math score in SAT of 52 
sat_2017.drop(sat_2017[sat_2017['Math'] == 52].index, inplace = True)

In [10]:
# Drop row that have a science score in ACT of 2.3 
act_2017.drop(act_2017[act_2017['Science'] == 2.3].index, inplace = True)

#### 5. What are your data types? 
Display the data types of each feature. 

In [11]:
sat_2017.dtypes

State                                 object
Participation                         object
Evidence-Based Reading and Writing     int64
Math                                   int64
Total                                  int64
dtype: object

In [12]:
act_2017.dtypes

State             object
Participation     object
English          float64
Math             float64
Reading          float64
Science          float64
Composite         object
dtype: object

What did you learn?
- Do any of them seem odd?  
- Which ones are not as they should be?  

##### Answer: 

- Participations in both datasets should be float.
- Composite is the average score of ACT of each state. However, the dtype is an object.


#### 6. Fix Incorrect Data Types
Based on what you discovered above, use appropriate methods to re-type incorrectly typed data.
- Define a function that will allow you to convert participation rates to an appropriate numeric type. Use `map` or `apply` to change these columns in each dataframe.

##### i. Participation

In [13]:
# Remove unneccessary character in participation column using function
def percent_to_float(prc):
    return int(prc.replace('%',''))/100

In [14]:
sat_2017["Participation"] = sat_2017["Participation"].apply(percent_to_float)

In [15]:
act_2017["Participation"] = act_2017["Participation"].apply(percent_to_float)

##### ii. Composite

- Fix any individual values preventing other columns from being the appropriate type.

In [16]:
# Find out what's wrong with composite column
act_2017.Composite.unique()

array(['21.0', '19.2', '19.8', '19.7', '19.4', '22.8', '20.8', '25.2',
       '24.1', '24.2', '21.4', '19.0', '22.3', '22.6', '21.9', '21.7',
       '20.0', '19.5', '24.3', '25.4', '21.5', '18.6', '20.4', '20.3',
       '17.8', '25.5', '23.9', '19.1', '22.0', '21.8', '23.7', '24.0',
       '18.7', '20.7', '23.6', '23.8', '20.5', '20.2x'], dtype=object)

There is a typo in one of the row. It should be 20.2 instead of 20.2x

In [17]:
# Remove typo 
act_2017.replace({'Composite': {"20.2x": "20.2"}}, inplace = True)
# Convert dtype to float
act_2017["Composite"] = act_2017["Composite"].astype(float) 

- Display the data types again to confirm they are correct.

In [18]:
sat_2017.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 50 entries, 0 to 50
Data columns (total 5 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   State                               50 non-null     object 
 1   Participation                       50 non-null     float64
 2   Evidence-Based Reading and Writing  50 non-null     int64  
 3   Math                                50 non-null     int64  
 4   Total                               50 non-null     int64  
dtypes: float64(1), int64(3), object(1)
memory usage: 2.3+ KB


In [19]:
act_2017.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 51 entries, 0 to 51
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   State          51 non-null     object 
 1   Participation  51 non-null     float64
 2   English        51 non-null     float64
 3   Math           51 non-null     float64
 4   Reading        51 non-null     float64
 5   Science        51 non-null     float64
 6   Composite      51 non-null     float64
dtypes: float64(6), object(1)
memory usage: 3.2+ KB


#### 7. Rename Columns
Change the names of the columns to more expressive names so that you can tell the difference the SAT columns and the ACT columns. Your solution should map all column names being changed at once (no repeated singular name-changes). **We will be combining these data with some of the data from 2018, and so you should name columns in an appropriate way**.

**Guidelines**:
- Column names should be all lowercase (you will thank yourself when you start pushing data to SQL later in the course)
- Column names should not contain spaces (underscores will suffice--this allows for using the `df.column_name` method to access columns in addition to `df['column_name']`.
- Column names should be unique and informative (the only feature that we actually share between dataframes is the state).

In [20]:
sat_2017.rename(columns={'State': 'state',
                         'Participation': 'sat_part',
                         'Evidence-Based Reading and Writing':'sat_erw',
                         'Math': 'sat_math',
                         'Total': 'sat_total'
                        }, inplace=True)

sat_2017.head()

state  sat_part  sat_erw  sat_math  sat_total
0     Alabama      0.05      593       572       1165
1      Alaska      0.38      547       533       1080
2     Arizona      0.30      563       553       1116
3    Arkansas      0.03      614       594       1208
4  California      0.53      531       524       1055

In [21]:
act_2017.rename(columns={'State': 'state',
                         'Participation': 'act_part',
                         'English':'act_eng',
                         'Math': 'act_math',
                         'Reading': 'act_read',
                         'Science': 'act_sci',
                         'Composite': 'act_composite'
                        }, inplace=True)

act_2017.head()

state  act_part  act_eng  act_math  act_read  act_sci  act_composite
0  National      0.60     20.3      20.7      21.4     21.0           21.0
1   Alabama      1.00     18.9      18.4      19.7     19.4           19.2
2    Alaska      0.65     18.7      19.8      20.4     19.9           19.8
3   Arizona      0.62     18.6      19.8      20.1     19.8           19.7
4  Arkansas      1.00     18.9      19.0      19.7     19.5           19.4

#### 8. Create a data dictionary

Now that we've fixed our data, and given it appropriate names, let's create a [data dictionary](http://library.ucmerced.edu/node/10249). 

A data dictionary provides a quick overview of features/variables/columns, alongside data types and descriptions. The more descriptive you can be, the more useful this document is.

Example of a Fictional Data Dictionary Entry: 

|Feature|Type|Dataset|Description|
|---|---|---|---|
|**county_pop**|*integer*|2010 census|The population of the county (units in thousands, where 2.5 represents 2500 people).| 
|**per_poverty**|*float*|2010 census|The percent of the county over the age of 18 living below the 200% of official US poverty rate (units percent to two decimal places 98.10 means 98.1%)|

[Here's a quick link to a short guide for formatting markdown in Jupyter notebooks](https://jupyter-notebook.readthedocs.io/en/stable/examples/Notebook/Working%20With%20Markdown%20Cells.html).

Provided is the skeleton for formatting a markdown table, with columns headers that will help you create a data dictionary to quickly summarize your data, as well as some examples. **This would be a great thing to copy and paste into your custom README for this project.**

|Feature|Type|Dataset|Description|
|---|---|---|---|
|column name|int/float/object|ACT/SAT|This is an example| 


|Feature|Type|Dataset|Description|
|---|---|---|---|
|state|*object*|final|Names of all US states and the District of Columbia| 
|sat_part_17 |*float*|final|Percentage of students participated in the SAT test in 2017|
|erw_17|*int*|SAT|final|Evidence-Based Reading and Writing score in 2017 |
|sat_math_17|*int*|final|SAT Math Score in 2017|
|sat_total_17|*int*|final|Sum of ERW and Math scores in 2017|
|act_part_17|*float*|final|The percentage of students participated in the ACT test|
|act_eng_17|*float*|final|ACT English Score in 2017|
|act_math_17|*float*|final|ACT Math Score in 2017|
|act_read_17|*float*|final|ACT Reading Score in 2017|
|act_sci_17|*float*|final|ACT Science Score in 2017|
|act_composite_17|*float*|final|The mean score across all components of the ACT in 2017|
|sat_part_18|*float*|final|Percentage of students participated in the SAT test in 2018|
|erw_18|*int*|final|Evidence-Based Reading and Writing score in 2018|
|sat_math_18|*int*|final|SAT Math Score in 2018|
|sat_total_18|*int*|final|Sum of ERW and Math scores in 2018|
|act_part_18|*float*|final|Percentage of students participated in the ACT test in 2018|
|act_eng_18|*float*|final|ACT English Score in 2018|
|act_math_18|*float*|final|ACT Math Score in 2018|
|act_read_18|*float*|final|ACT Reading Score in 2018|
|act_sci_18|*float*|final|ACT Science Score in 2018|
|act_composite_18|*float*|final|The mean score across all components of the ACT in 2018|

#### 9. Drop unnecessary rows

One of our dataframes contains an extra row. Identify and remove this from the dataframe.

In [22]:
act_2017.drop([0], inplace= True)
act_2017.head()

state  act_part  act_eng  act_math  act_read  act_sci  act_composite
1     Alabama      1.00     18.9      18.4      19.7     19.4           19.2
2      Alaska      0.65     18.7      19.8      20.4     19.9           19.8
3     Arizona      0.62     18.6      19.8      20.1     19.8           19.7
4    Arkansas      1.00     18.9      19.0      19.7     19.5           19.4
5  California      0.31     22.5      22.7      23.1     22.2           22.8

#### 10. Merge Dataframes

Join the 2017 ACT and SAT dataframes using the state in each dataframe as the key. Assign this to a new variable.

In [23]:
combined_2017 = pd.merge(sat_2017,act_2017, on = "state")
combined_2017.head()

state  sat_part  sat_erw  sat_math  sat_total  act_part  act_eng  \
0     Alabama      0.05      593       572       1165      1.00     18.9   
1      Alaska      0.38      547       533       1080      0.65     18.7   
2     Arizona      0.30      563       553       1116      0.62     18.6   
3    Arkansas      0.03      614       594       1208      1.00     18.9   
4  California      0.53      531       524       1055      0.31     22.5   

   act_math  act_read  act_sci  act_composite  
0      18.4      19.7     19.4           19.2  
1      19.8      20.4     19.9           19.8  
2      19.8      20.1     19.8           19.7  
3      19.0      19.7     19.5           19.4  
4      22.7      23.1     22.2           22.8

#### 11. Save your cleaned, merged dataframe

Use a relative path to save out your data as `combined_2017.csv`.

In [24]:
#code
combined_2017.to_csv('../dataset/combined_2017.csv')

## 2018 Data Import and Cleaning

Links to the 2018 ACT and SAT data are provided in the README. These data live in PDFs, and so you'll get to enjoy practicing some *manual* data collection. Save these data as a CSV in your `data` directory, and import, explore, and clean these data in the same way you did above. **Make sure you comment on your steps so it is clear *why* you are doing each process**.

#### 1. Load and display the CSV files into DataFrames

In [25]:
sat_2018 = pd.read_csv('../dataset/sat_2018.csv')
sat_2018.head()

State Participation  Evidence-Based Reading and Writing  Math  Total
0     Alabama            6%                                 595   571   1166
1      Alaska           43%                                 562   544   1106
2     Arizona           29%                                 577   572   1149
3    Arkansas            5%                                 592   576   1169
4  California           60%                                 540   536   1076

In [26]:
# act_2018 = pd.read_csv('../dataset/act_2018.csv')
# act_2018.head()

In [27]:
act_2018 = pd.read_csv("../dataset/act_2018_updated.csv")
act_2018.head()

State  Percentage of Students Tested  Average Composite Score  \
0     Alabama                            100                     19.1   
1      Alaska                             33                     20.8   
2     Arizona                             66                     19.2   
3    Arkansas                            100                     19.4   
4  California                             27                     22.7   

   Average English Score  Average Math Score  Average Reading Score  \
0                   18.9                18.3                   19.6   
1                   19.8                20.6                   21.6   
2                   18.2                19.4                   19.5   
3                   19.1                18.9                   19.7   
4                   22.5                22.5                   23.0   

   Average Science Score  
0                   19.0  
1                   20.7  
2                   19.2  
3                   19.4  
4                   22.1

I have chosen to import the act_2018_updated.csv file as its data is complete and is identical to the columns specified in act_2017.

#### 2. Check if there is any obvious issues

In [28]:
sat_2018.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 5 columns):
 #   Column                              Non-Null Count  Dtype 
---  ------                              --------------  ----- 
 0   State                               51 non-null     object
 1   Participation                       51 non-null     object
 2   Evidence-Based Reading and Writing  51 non-null     int64 
 3   Math                                51 non-null     int64 
 4   Total                               51 non-null     int64 
dtypes: int64(3), object(2)
memory usage: 2.1+ KB


In [29]:
act_2018.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 7 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   State                          51 non-null     object 
 1   Percentage of Students Tested  51 non-null     int64  
 2   Average Composite Score        51 non-null     float64
 3   Average English Score          51 non-null     float64
 4   Average Math Score             51 non-null     float64
 5   Average Reading Score          51 non-null     float64
 6   Average Science Score          51 non-null     float64
dtypes: float64(5), int64(1), object(1)
memory usage: 2.9+ KB


#### 3. Cleaning Columns from unneccesary characters and incorrect dtypes

In [30]:
# Remove percentage sign
sat_2018["Participation"] = sat_2018["Participation"].apply(percent_to_float)
sat_2018.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 5 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   State                               51 non-null     object 
 1   Participation                       51 non-null     float64
 2   Evidence-Based Reading and Writing  51 non-null     int64  
 3   Math                                51 non-null     int64  
 4   Total                               51 non-null     int64  
dtypes: float64(1), int64(3), object(1)
memory usage: 2.1+ KB


In [31]:
# Create function to change percentage column from int to float
def to_float(num):
    return float(num/100)

In [32]:
act_2018["Percentage of Students Tested"] = act_2018["Percentage of Students Tested"].apply(to_float)
act_2018.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 7 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   State                          51 non-null     object 
 1   Percentage of Students Tested  51 non-null     float64
 2   Average Composite Score        51 non-null     float64
 3   Average English Score          51 non-null     float64
 4   Average Math Score             51 non-null     float64
 5   Average Reading Score          51 non-null     float64
 6   Average Science Score          51 non-null     float64
dtypes: float64(6), object(1)
memory usage: 2.9+ KB


#### 4. Rename Columns

In [33]:
sat_2018.rename(columns ={'State': 'state',
                          'Participation':'sat_part',
                          'Evidence-Based Reading and Writing':'sat_erw',
                          'Math': 'sat_math',
                          'Total': 'sat_total'
                         }, inplace=True)

sat_2018.head()

state  sat_part  sat_erw  sat_math  sat_total
0     Alabama      0.06      595       571       1166
1      Alaska      0.43      562       544       1106
2     Arizona      0.29      577       572       1149
3    Arkansas      0.05      592       576       1169
4  California      0.60      540       536       1076

In [34]:
act_2018.rename(columns={"State": "state", 
                         "Percentage of Students Tested": "act_part",
                         "Average Composite Score": "act_composite",
                         "Average English Score": "act_eng",
                         "Average Math Score": "act_math",
                         "Average Reading Score": "act_read",
                         "Average Science Score": "act_sci"
                        }, inplace=True)

act_2018.head()

state  act_part  act_composite  act_eng  act_math  act_read  act_sci
0     Alabama      1.00           19.1     18.9      18.3      19.6     19.0
1      Alaska      0.33           20.8     19.8      20.6      21.6     20.7
2     Arizona      0.66           19.2     18.2      19.4      19.5     19.2
3    Arkansas      1.00           19.4     19.1      18.9      19.7     19.4
4  California      0.27           22.7     22.5      22.5      23.0     22.1

In [35]:
# Rearrange Columns 
act_2018 = act_2018[["state", 
                     "act_part", 
                     "act_eng", 
                     "act_math", 
                     "act_read", 
                     "act_sci", 
                     "act_composite"
                    ]]
act_2018.head()

state  act_part  act_eng  act_math  act_read  act_sci  act_composite
0     Alabama      1.00     18.9      18.3      19.6     19.0           19.1
1      Alaska      0.33     19.8      20.6      21.6     20.7           20.8
2     Arizona      0.66     18.2      19.4      19.5     19.2           19.2
3    Arkansas      1.00     19.1      18.9      19.7     19.4           19.4
4  California      0.27     22.5      22.5      23.0     22.1           22.7

In [36]:
# Merge SAT and ACT 2018
combined_2018 = pd.merge(sat_2018, act_2018, on="state")
combined_2018.head()

state  sat_part  sat_erw  sat_math  sat_total  act_part  act_eng  \
0     Alabama      0.06      595       571       1166      1.00     18.9   
1      Alaska      0.43      562       544       1106      0.33     19.8   
2     Arizona      0.29      577       572       1149      0.66     18.2   
3    Arkansas      0.05      592       576       1169      1.00     19.1   
4  California      0.60      540       536       1076      0.27     22.5   

   act_math  act_read  act_sci  act_composite  
0      18.3      19.6     19.0           19.1  
1      20.6      21.6     20.7           20.8  
2      19.4      19.5     19.2           19.2  
3      18.9      19.7     19.4           19.4  
4      22.5      23.0     22.1           22.7

In [37]:
#Save file
combined_2018.to_csv('../dataset/combined_2018.csv')

#### Combine your 2017 and 2018 data into a single dataframe
Joining on state names should work, assuming you formatted all your state names identically. Make sure none of your columns (other than state) have identical names. Do yourself a favor and decide if you're encoding participation rates as floats or integers and standardize this across your datasets.

Save the contents of this merged dataframe as `final.csv`.

**Use this combined dataframe for the remainder of the project**.

In [38]:
#Merge 2017 and 2018 Dataframes
final = pd.merge(combined_2017, combined_2018, on="state", suffixes = ["_17","_18"])
final.head()

state  sat_part_17  sat_erw_17  sat_math_17  sat_total_17  \
0     Alabama         0.05         593          572          1165   
1      Alaska         0.38         547          533          1080   
2     Arizona         0.30         563          553          1116   
3    Arkansas         0.03         614          594          1208   
4  California         0.53         531          524          1055   

   act_part_17  act_eng_17  act_math_17  act_read_17  act_sci_17  ...  \
0         1.00        18.9         18.4         19.7        19.4  ...   
1         0.65        18.7         19.8         20.4        19.9  ...   
2         0.62        18.6         19.8         20.1        19.8  ...   
3         1.00        18.9         19.0         19.7        19.5  ...   
4         0.31        22.5         22.7         23.1        22.2  ...   

   sat_part_18  sat_erw_18  sat_math_18  sat_total_18  act_part_18  \
0         0.06         595          571          1166         1.00   
1         0.43         562          544          1106         0.33   
2         0.29         577          572          1149         0.66   
3         0.05         592          576          1169         1.00   
4         0.60         540          536          1076         0.27   

   act_eng_18  act_math_18  act_read_18  act_sci_18  act_composite_18  
0        18.9         18.3         19.6        19.0              19.1  
1        19.8         20.6         21.6        20.7              20.8  
2        18.2         19.4         19.5        19.2              19.2  
3        19.1         18.9         19.7        19.4              19.4  
4        22.5         22.5         23.0        22.1              22.7  

[5 rows x 21 columns]

In [39]:
#Saving File into Data folder
final.to_csv('../dataset/final.csv', index = False)

----> Proceed to the next notebook for [EDA](./02_Exploratory_Data_Analysis.ipynb)